packages/typespec-vscode/src/openapi3-preview.ts

```diff
@@ -0,0 +1,251 @@
+    watch.onDidCreate(throttledChangeHandler);
+    watch.onDidDelete(throttledChangeHandler);
+
+    panel.onDidDispose(() => {
+      openApi3PreviewPanels.delete(mainTspFile);
+      selectedOpenApi3OutputFiles.delete(mainTspFile);
+      watch.dispose();
+    });
+
+    panel.webview.onDidReceiveMessage(async (message) => {
+      switch (message.event) {
+        case "initialized":
+          void panel.webview.postMessage({ command: "load", param: fileContent });
+          break;
+        default:
+          logger.error(`Unknown WebView event received: ${message}`);
+          break;
+      }
+    });
+
+    loadHtml(context.extensionUri, panel);
+  }
+}
+
+let html: string | undefined;
+
+function loadHtml(extensionUri: vscode.Uri, panel: vscode.WebviewPanel) {
+  if (html === undefined) {
+    const swaggerUiRoot = vscode.Uri.joinPath(extensionUri, "swagger-ui");
+    const css = panel.webview.asWebviewUri(vscode.Uri.joinPath(swaggerUiRoot, "swagger-ui.css"));
+    const bundleJs = panel.webview.asWebviewUri(
+      vscode.Uri.joinPath(swaggerUiRoot, "swagger-ui-bundle.js"),
+    );
+
+    const presetJs = panel.webview.asWebviewUri(
+      vscode.Uri.joinPath(swaggerUiRoot, "swagger-ui-standalone-preset.js"),
+    );
+    // initialization script is customized to load openapi3 spec from openapi emitter output
+    const initJs = panel.webview.asWebviewUri(
+      vscode.Uri.joinPath(swaggerUiRoot, "swagger-initializer.js"),
+    );
+
+    html = `<!doctype html>
+            <html lang="en">
+              <head>
+                <meta charset="utf-8">
+                <meta name="viewport" content="width=device-width, initial-scale=1">
+                <link rel="stylesheet" type="text/css" href="${css}" />
+              </head>
+              <body class="no-margin-padding full-device-view" style="background-color:white">
+                <div id="swagger-ui"></div>
+                <script src="${bundleJs}" charset="UTF-8"> </script>
+                <script src="${presetJs}" charset="UTF-8"> </script>
+                <script src="${initJs}" charset="UTF-8"> </script>
+              </body>
+            </html>`;
+  }
+
+  panel.webview.html = html;
+}
+
+/**
+ * Throttle the function to be called at most once in every blockInMs milliseconds. This utility
+ * is useful when your event handler will trigger the same event multiple times in a short period.
+ *
+ * @param fn Underlying function to be throttled
+ * @param blockInMs Block time in milliseconds
+ * @returns a throttled function
+ */
+function throttle<T extends (...args: any[]) => any>(fn: T, blockInMs: number): T {
+  let time: number | undefined;
+  return function (this: any, ...args: Parameters<T>) {
+    const now = Date.now();
+    if (time === undefined || now - time >= blockInMs) {
+      time = now;
+      fn.apply(this, args);
+    }
+  } as T;
+}
+
+async function parseOpenApi3File(filePath: string): Promise<string | undefined> {
```

Comment:
make the func general and move to util?

Reply:
not sure how general this could be, but I've moved it into `utils.ts` first